subsys/bluetooth/mesh/net.c

```diff
@@ -695,27 +715,41 @@ static void bt_mesh_net_relay(struct net_buf_simple *sbuf,
 	sbuf->data[1] &= 0x80;
 	sbuf->data[1] |= rx->ctx.recv_ttl - 1U;
 
-	net_buf_add_mem(buf, sbuf->data, sbuf->len);
-
 	cred = &rx->sub->keys[SUBNET_KEY_TX_IDX(rx->sub)].msg;
 
-	BT_DBG("Relaying packet. TTL is now %u", TTL(buf->data));
+	BT_DBG("Relaying packet. TTL is now %u", TTL(sbuf->data));
 
 	/* Update NID if RX or RX was with friend credentials */
 	if (rx->friend_cred) {
-		buf->data[0] &= 0x80; /* Clear everything except IVI */
-		buf->data[0] |= cred->nid;
+		sbuf->data[0] &= 0x80; /* Clear everything except IVI */
+		sbuf->data[0] |= cred->nid;
 	}
 
 	/* We re-encrypt and obfuscate using the received IVI rather than
 	 * the normal TX IVI (which may be different) since the transport
 	 * layer nonce includes the IVI.
 	 */
-	if (net_encrypt(&buf->b, cred, BT_MESH_NET_IVI_RX(rx), false)) {
+	net_buf_add_u8(buf, sbuf->data[0]);
+	net_buf_add(buf, sbuf->len - 1 + BT_MESH_NET_MIC_LEN(sbuf->data));
+
+	err = bt_mesh_net_encrypt(cred->enc, &sbuf->data[NET_PDU_ENC_OFFSET],
+				  sbuf->len - NET_PDU_ENC_OFFSET,
+				  sbuf->data, &buf->data[NET_PDU_ENC_OFFSET],
+				  BT_MESH_NET_MIC_LEN(sbuf->data),
+				  BT_MESH_NET_IVI_RX(rx), false);
+	if (err) {
 		BT_ERR("Re-encrypting failed");
 		goto done;
 	}
 
+	err = bt_mesh_net_obfuscate(sbuf->data, buf->data,
```

Comment:
I suppose this is not correct to perform obfuscation from one buffer to another buffer. 
Actually, having input and output parameters doesn't have a sense for obfuscation. Obfuscation is done for the frame and it is inserted into the same frame without going into third resources.

Reply:
This is the basis of the execution of this PR, and I don't want to address part of it because it doesn't make much sense.

Reply:
Hm. I'm not sure in that this trick is required to remove extra copying. What does prevent to create outgoing frame and obfuscate it from its own memory rather than from buffer of incoming frame?